### PR TITLE
Traits concept

### PR DESCRIPTION
**Pull request by lijas**
How I think traits could work. Note this does not work at the moment, it is only concept. .

Benefits:
The user dont have to modify the element routine depending on material type, but can get the material response from all materials (by passing the deformation gradient).

Some other stuff I though about:
I think we should decide what the material should output, ie dSdE or dPdF etc...
If we should allow "delta-strains" (Δε) or force ε + state variable.

### CODE REVIEW

**Comment by lijas** [comment]
So the goal with traits is to be able to call any material function from your element routine and get back the stress and tangent that you element routine is based upon.

So if your element routine is based on the work pairs S and E, you can now call `material_response_S_dSdE` (name up for discussion), and you will get back S and dSdE.

If someone else uses element routine with work pair P and dPdF, they can call material_response_P_dPdF.

and so on... People can add more work pairs function if they like, but I guess there are only 2 or 3 common ones...

Instead of having very explicit function names as above, it could be possible to also associate a material with a "output" trait, and then we could maybe reduce the code a bit (now it quite a lot of copy pasting), but I dont think we should overuse the trait system...

**Comment by kimauth** [comment]
What do you think of allowing to pass a singleton type, similar to how the stress/strain wrappers work? The default could be interfered from the `StrainTrait` (to be the work-conjugate stress type to the strain type probably), but in an element routine that requires a certain stress type the singleton type could be passed to make sure to get the right returns.

I think this wouldn't really be any different to the explicit function names, but it would make the call to the material routine much more consistent with the small strain routines and it would not add an additional trait (also the returned stress type should probably not be a property of the material - you could want the same material to return S in one element routine and P in another).

File: src/traits.jl [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    Pᵀ, dPᵀdF, state = material_response(m, F, state, Δt, cahce=cache, options=options)
+    return Pᵀ, dPᵀdF, state
+end
+
+#
+# RightCauchyGreen
+function material_response_Pᵀ_dPᵀdF(::RightCauchyGreen, m::AbstractMaterial, F::Tensor{2}, state::AbstractMaterialState, Δt::Float64; cache, options)
+    C = tdot(F)
+    S, dSdC, state = material_response(m, C, state, Δt, cahce=cache, options=options)
+    dPᵀdF =  otimesu(F,I) ⊡ dSdE ⊡ otimesu(F',I) + otimesu(S,I)
```
[the remark this answers is not in the document]
Comment (koehlerson) [reply src/traits.jl:64]:
```suggestion
    dPᵀdF =  otimesu(F,I) ⊡ dSdC ⊡ otimesu(F',I) + otimesu(S,I)
```

File: src/traits.jl [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    return S, dSdE, state
+end
+
+#
+# DeformationGradient
+function material_response_S_dSdE(::DeformationGradient, m::AbstractMaterial, F::Tensor{2}, state::AbstractMaterialState, Δt::Float64; cache, options)
+    Pᵀ, dPᵀdF, state = material_response(m, F, state, Δt, cahce=cache, options=options)
+    S = Pᵀ ⋅ inv(F')
+    dSdE = 2 * inv( otimesu(F,I) ) ⊡ (dPᵀdF - otimesu(S,I)) ⊡ inv( otimesu(F',I) )
+    return S, 2dSdC, state
```
[the remark this answers is not in the document]
Comment (koehlerson) [reply src/traits.jl:40]:
```suggestion
    return S, 2dSdE, state 
```

**Review by koehlerson — COMMENTED** [review]
I know its WIP, but just some notes, if you want to merge it at some point

**Comment by lijas** [comment]
Thanks @koehlerson 

I was able to slim the code a bit by defing a "tangent output struct/trait"

The user call is now:
```
mat = NeoHook(λ = 1.0, µ = 1.0)
state = initial_material_state(mat)
F = Tensor((2.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0))
Δt = 0.0

S, dSdE, state = material_response(dSdE(), mat, F, state, Δt)
```

File: src/traits.jl [hunk 3]
```diff
@@ -6,66 +7,56 @@ struct VelocityGradient <: StrainMeasure end
...
-    material_response_Pᵀ_dPᵀdF(strainmeasure(M), m, F, state, Δt, cache, options)
-end
 
-#
-# DeformationGradient
-function material_response_Pᵀ_dPᵀdF(::DeformationGradient, m::AbstractMaterial, F::Tensor{2}, state::AbstractMaterialState, Δt::Float64; cache, options)
-    Pᵀ, dPᵀdF, state = material_response(m, F, state, Δt, cahce=cache, options=options)
-    return Pᵀ, dPᵀdF, state
+function material_response(output_tangent::AbstractTangent, m::AbstractMaterial, F::Tensor{2}, state::AbstractMaterialState, Δt::Float64 = 0.0; cache=nothing, options=nothing)
+    return material_response(output_tangent, strainmeasure(m), m, F, state, Δt, cahce=cache, options=options)
```
Comment (koehlerson) [inline src/traits.jl:52]:
```suggestion
    return material_response(output_tangent, strainmeasure(m), m, F, state, Δt, cache=cache, options=options)
```

File: src/traits.jl [hunk 4]
```diff
@@ -6,66 +7,56 @@ struct VelocityGradient <: StrainMeasure end
...
-    C = tdot(F)
-    S, dSdC, state = material_response(m, C, state, Δt, cahce=cache, options=options)
-    dPᵀdF =  otimesu(F,I) ⊡ dSdE ⊡ otimesu(F',I) + otimesu(S,I)
-    Pᵀ = S⋅F'
-    return Pᵀ, dPᵀdF, state
-end
+function material_response(output_tangent::AbstractTangent, straintype::StrainMeasure, m::AbstractMaterial, F::Tensor{2}, state::AbstractMaterialState, Δt::Float64; cache, options)
+    
+    strain = compute_strain(F, straintype)
+    stress, strain, newstate = material_response(m, strain, state, Δt, cahce=cache, options=options)
```
Comment (koehlerson) [inline src/traits.jl:58]:
```suggestion
    stress, strain, newstate = material_response(m, strain, state, Δt, cache=cache, options=options)
```

**Review by koehlerson — COMMENTED** [review]
thats nice! Also nice to collect the material tangent transformation in this repo!

**Comment by lijas** [comment]
Annoyingly this gives an error
`S, dSdE, state = material_response(dSdE(), mat, F, state, Δt)`
I must change name of the output variable name `dSdE` to something else for it work. Should find a new name for `dSdE()`...maybe with partials deriviatives instead.

**Comment by lijas** [comment]
I suck at git.. I did a rebase and all commits duplicated for some reason... Does not matter though, it will be squashed-merged.

**Comment by lijas** [comment]
I messed this branch up...open new PR...

**Pull request closed without merging**